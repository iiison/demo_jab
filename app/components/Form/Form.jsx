import React, { createContext } from 'react'
import { PropTypes } from 'prop-types'

import { useSetState } from '$UTILS/reactUtils'
import validations from './validationRules'

export const FormCtx = createContext({})

export function useFormCTXProvider() {
  const [data, setData] = useSetState()

  return {
    data,
    setData,
    FormCTXProvider : FormCtx.Provider,
  }
}

function validateRule({ id, value, displayName, ruleValue, ruleArgs, validation }) {
  if (validation) {
    const stringifiedValue = value.toString()
    const isRuleSatisfied = ( ruleValue !== 'required' && !stringifiedValue )
      ? true
      : validation.rule.apply(null, ruleArgs).test(value.toString())

    let error = ''

    if (!isRuleSatisfied) {
      error = validation.formatter.apply(null, [displayName || id, ...ruleArgs])

      return error
    }
  } else {
    throw `invalid validation rule: ${ruleValue}, please use an existing validation rule name or pass a custom function with same name through 'customRules' prop in Input: ${id}. Rule value should be an object with keys: 'rule' as an Regex and 'formatter' as a function, that formats the value.` // eslint-disable-line
  }

  return ''
}

export function validateField({
  id,
  value,
  validate,
  displayName,
  customRules = {}
}) {
  let error = ''
  const rules = validate ? validate.split('|') : ''

  if (rules.length > 0) {
    for (const rule of rules) {
      if (!rule) {
        continue
      }

      const ruleDetails = rule.split('-')
      const [ruleValue, ...ruleArgs] = ruleDetails
      const validation = validations[ruleValue] || customRules[ruleValue]

      try {
        error = validateRule({ id, value, displayName, ruleValue, ruleArgs, validation })

        if (error !== '') {
          break
        }
      } catch (invalidRuleError) {
        throw invalidRuleError
      }
    }
  }

  return error
}

export function useUpdateField(setData, isNewField) {
  return (field, shouldRunValidation = true) => {
    const { id, isPristine, error } = field
    let validationError = error

    if (id === undefined) {
      throw new Error(`Please pass id to ${JSON.stringify(field)}`)
    }

    const value = isNewField === true && field.value === undefined ? '' : field.value

    if (isNewField !== true && isPristine === false && shouldRunValidation === true) {
      validationError = validateField(field)
    }

    setData({
      [id] : {
        ...field,
        value,
        error : validationError
      }
    })

  }
}

export default function Form({ children }) {
  const {
    data,
    setData,
    FormCTXProvider
  } = useFormCTXProvider()
  const addField = useUpdateField(setData, true)
  const updateField = useUpdateField(setData, false)
  const value = {
    data,
    addField,
    updateField,
    validateField
  }

  return (
    <FormCTXProvider value={value}>
      <form action="" className='col-12 grid'>
        {children}
      </form>
    </FormCTXProvider>
  )
}

Form.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
}

