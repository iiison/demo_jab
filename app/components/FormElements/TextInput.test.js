import React, { useContext, createContext } from 'react'
import { renderHook, act } from '@testing-library/react-hooks'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Form from '../Form/Form'
import Input, {
  useHandleBlur,
  useHandleChange,
} from './TextInput'

const TestCtx = createContext()

describe('>> COMPONENT -- TextInput Tests', () => {
  const field = {
    id : 'test',
    value : 'some value'
  }
  
  describe('• HOOK - useHandleChange Tests', () => {
    it('checks if hook returns function', () => {
      const updateField = jest.fn()
      const handleChange = useHandleChange(updateField, field)

      expect(typeof handleChange).toBe('function')
    })

    it('handleChange triggers updateField when called', () => {
      const updateField = jest.fn()
      const handleChange = useHandleChange(updateField, field)
      const event = {
        persist : jest.fn(),
        target  : {
          value : 'some value'
        }
      }

      handleChange(event)
      expect(event.persist).toHaveBeenCalled()
      expect(updateField).toHaveBeenCalledWith({
        ...field,
        value : event.target.value
      })
    })

    it('handleChange changes `isPristine` field to "partial" if it is set to true', () => {
      const fieldCopy = {
        ...field,
        isPristine : true
      }
      const updateField = jest.fn()
      const handleChange = useHandleChange(updateField, fieldCopy)
      const event = {
        persist : jest.fn(),
        target  : {
          value : 'some value'
        }
      }

      handleChange(event)
      expect(updateField).toHaveBeenCalledWith({
        ...fieldCopy,
        value : event.target.value,
        isPristine : 'partial'
      })
    })

    it('handleChange calls userdefined onChange method from field', () => {
      const fieldCopy = {
        ...field,
        events : {
          onChange : jest.fn()
        }
      }
      const updateField = jest.fn()
      const handleChange = useHandleChange(updateField, fieldCopy)
      const event = {
        persist : jest.fn(),
        target  : {
          value : 'some value'
        }
      }

      handleChange(event)
      expect(fieldCopy.events.onChange).toHaveBeenCalledWith({
        ...fieldCopy,
        value : event.target.value
      })
    })
  })

  describe('• HOOK - useHandleBlur Tests', () => {
    it('checks if hook returns function', () => {
      const updateField = jest.fn()
      const validateField = jest.fn()

      const handleBlur = useHandleBlur(updateField, validateField, field)

      expect(typeof handleBlur).toBe('function')
    })

    it('handleBlur should not call updateField and validateField if isPristine set to true', () => {
      const fieldCopy = {
        ...field,
        isPristine : true
      }
      const updateField = jest.fn()
      const validateField = jest.fn()

      const handleBlur = useHandleBlur(updateField, validateField, fieldCopy)
      const event = {
        persist : jest.fn()
      }

      handleBlur(event)
      expect(event.persist).toHaveBeenCalled()
      expect(updateField).not.toHaveBeenCalled()
      expect(validateField).not.toHaveBeenCalled()
    })

    it ('triggers updateField and validateField function when isPristine is set to "partial"', () => {
      const fieldCopy = {
        ...field,
        isPristine : 'partial'
      }
      const updateField = jest.fn()
      const validateField = jest.fn()
      const handleBlur = useHandleBlur(updateField, validateField, fieldCopy)
      const event = {
        persist : jest.fn()
      }

      handleBlur(event)
      expect(event.persist).toHaveBeenCalled()
      expect(validateField).toHaveBeenCalledWith({
        ...fieldCopy,
        isPristine : false
      })
      expect(updateField).toHaveBeenCalledWith({
        ...fieldCopy,
        isPristine : false,
        error : undefined
      }, false)
    })

    it ('triggers updateField and validateField function when isPristine is set to false', () => {
      const fieldCopy = {
        ...field,
        isPristine : false
      }
      const updateField = jest.fn()
      const validateField = jest.fn()
      const handleBlur = useHandleBlur(updateField, validateField, fieldCopy)
      const event = {
        persist : jest.fn()
      }

      handleBlur(event)
      expect(event.persist).toHaveBeenCalled()
      expect(validateField).toHaveBeenCalledWith({
        ...fieldCopy,
        isPristine : false
      })
      expect(updateField).toHaveBeenCalledWith({
        ...fieldCopy,
        isPristine : false,
        error : undefined
      }, false)
    })

    it ('triggers userdefined onBlur function', () => {
      const fieldCopy = {
        ...field,
        events : {
          onBlur : jest.fn()
        }
      }
      const updateField = jest.fn()
      const validateField = jest.fn()
      const handleBlur = useHandleBlur(updateField, validateField, fieldCopy)
      const event = {
        persist : jest.fn()
      }

      handleBlur(event)
      expect(fieldCopy.events.onBlur).toHaveBeenCalledWith({ ...fieldCopy })
    })
  })

  describe('• Component - UI Tests', () => {
    it('renders the Input component snapshot', () => {
      const { container } = render(<Form><Input {...field} /></Form>)

      expect(container).toMatchSnapshot()
    })

    it ('triggers onChange on input field', () => {
      const fieldCopy = {
        ...field,
        placeholder : 'test',
        events : {
          onChange : jest.fn()
        }
      }
     
      const data = render(
        <Form>
          <Input {...fieldCopy} />
        </Form>
      )
      const input = data.getByPlaceholderText('test')

      fireEvent.change(input, { target : { value : 'test' }})

      expect(input.value).toEqual('test')
    })

    it ('triggers onBlur on input field', () => {
      const fieldCopy = {
        ...field,
        placeholder : 'test',
        events : {
          onBlur : jest.fn()
        }
      }
     
      const data = render(
        <Form>
          <Input {...fieldCopy} />
        </Form>
      )
      const input = data.getByPlaceholderText('test')

      input.focus()
      input.blur()

      expect(fieldCopy.events.onBlur).toHaveBeenCalled()
    })
  })
})
