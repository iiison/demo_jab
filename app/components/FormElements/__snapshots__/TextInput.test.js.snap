// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>> COMPONENT -- TextInput Tests â€¢ Component - UI Tests renders the Input component snapshot 1`] = `
<div>
  <form
    action=""
    class="col-12 grid"
  >
    <div
      class="grid col-12"
    >
      <label
        class="relative col label"
        for="test"
      >
        <div
          class="grid-middle"
        >
          <div
            class="col grid"
          >
            <input
              class="col-12 input"
              id="test"
              type="text"
              value="some value"
            />
          </div>
        </div>
      </label>
    </div>
  </form>
</div>
`;
